sel4-ftrace-merge/include/ftrace.h

```diff
@@ -0,0 +1,213 @@
+		struct ftrace_event_formats *ftrace_event_formats;
+	} ftrace_event_formats;
+	struct event_formats_wrap {
+		size_t size;
+		struct event_formats *event_formats;
+	} event_formats;
+	struct kallsyms_info_wrap {
+		size_t size;
+		struct kallsyms_info *kallsyms_info;
+	} kallsyms_info;
+	struct printk_info_wrap {
+		size_t size;
+		struct printk_info *printk_info;
+	} printk_info;
+	struct process_info_wrap {
+		size_t size;
+		struct process_info *process_info;
+	} process_info;
+	struct header_end_wrap {
+		size_t size;
+		struct header_end *header_end;
+	} header_end;
+	struct flyrecord_wrap {
+		size_t size;
+		struct flyrecord *flyrecords;
+	} flyrecords;
+	struct padding_wrap {
+		size_t size;
+		char *data;
+	} padding;
+	struct rest_data_wrap {
+		size_t size;
+		char *data;
+	} rest_data;
+};
+
+struct ftrace_initial {
+	uint8_t magic[3];
+	union {
+		struct {
+			char tracing[7];
+			char version[1];
+			char nullt[1];
+		} __attribute__((packed)) ver_dec;
+		char ver_str[7 + 1 + 1];
+	};
+	uint8_t endianess;
+	uint8_t bwidth;
+	uint32_t pagesize;
+} __attribute__((packed));
+
+struct ftrace_header_page_format {
+	char magic[12];
+	uint64_t size;
+
+	char data[];
+} __attribute__((packed));
+
+struct ftrace_header_event_format {
+	char magic[13];
+	uint64_t size;
+
+	char data[];
+} __attribute__((packed));
+
+struct ftrace_event_format {
+		uint64_t size;
+
+		char data[];
+} __attribute__((packed));
+
+struct ftrace_event_formats {
+	uint32_t count;
+
+	char data[];
+} __attribute__((packed));
+
+struct event_formats {
+	uint32_t count;
+
+	char data[];
+} __attribute__((packed));
+
+struct event_format_nxt {
+		uint64_t size;
+		char data[];
+} __attribute__((packed)) header;
+
+struct event_formats_nxt {
+	uint32_t count;
+
+	char data[];
+} __attribute__((packed));
+
+struct kallsyms_info {
+	uint32_t size;
+
+	char data[];
+} __attribute__((packed));
+
+struct printk_info {
+	uint32_t size;
+
+	char data[];
+} __attribute__((packed));
+
+struct process_info {
+	uint64_t size;
+
+	char data[];
+} __attribute__((packed));
+
+struct cpu_info {
+	uint32_t size;
+
+	struct {
+		uint64_t offset;
+		uint64_t size;
+	} __attribute__((packed)) data[];
+} __attribute__((packed));
+
+struct option_data {
+	uint32_t size;
+
+	char data[];
+} __attribute__((packed));
+
+struct option {
+	uint16_t option;
+
+	char data[];
+};
+
+struct command {
+	char name[10];
+} __attribute__((packed));
+
+struct header_end {
+	uint32_t n_cpu;
+
+	char data[];
+} __attribute__((packed));
+
+struct ftrace_func_field {
+	unsigned short common_type;
+	unsigned char common_flags;
+	unsigned char common_preempt_count;
+	int common_pid;
+
+	unsigned long ip;
+	unsigned long parent_ip;
+} __attribute__((packed));
+
+struct page_header {
+	u64 timestamp;
+	union {
+		local_t commit;
+		int overwrite;
+	};
+	char data[4080];
+} __attribute__((packed));
+
+struct flyrecord {
+	uint64_t offset;
+	uint64_t size;
+} __attribute__((packed));
+
+#endif // FTRACE_H
```

Comment:
Also why I started my rant about `#pragma once` (or lack of it) is that I see this closing #endif sometimes prefixed with old style C comment and sometimes with C++ style comment like this. Outside Linux kernel both work with default compiler C dialect (C11), but I think we should strive for consistency, choose one style and stick to it. I was thinking `#pragma once` would be easiest solution, no need to decide anything.

Reply:
Fixed comment to 'C dialect'.